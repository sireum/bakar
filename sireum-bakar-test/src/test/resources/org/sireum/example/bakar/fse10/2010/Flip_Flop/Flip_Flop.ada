package Flip_Flop
  --# own State: Bool;
  --# initializes State;
is
   --  The following two declarations are necessary only because Spark does
   --  not allow own variables of concrete types to be refined. Hopefully this
   --  can be added soon.
   --
   --# type Bool is abstract;
   --# function B(X: Bool) return Boolean;

   function Read return Boolean;
   --# global in State;
   --# return B(State);

   procedure Flip;
   --# global in out State;
   --# derives State from State;
   --# post B(State) = not B(State~);

end Flip_Flop;

package body Flip_Flop
  --# own State is N;
is
   N: Natural;

   --# function Represents(N: Natural) return Boolean;
   --! return not (N = 0);

   function Read return Boolean
     --# global in N;
     --# return not (N = 0);
   is begin
      return not (N = 0);
   end Read;

   procedure Flip
     --# global in out N;
     --# derives N from N;
     --# post (N~ = 0 -> N = 1) and
     --#      (N~ /= 0 -> N = 0);
   is begin
      if (N = 0) then
         N := 1;
      else
         N := 0;
      end if;
   end Flip;
begin
   N := 0;
end Flip_Flop;
