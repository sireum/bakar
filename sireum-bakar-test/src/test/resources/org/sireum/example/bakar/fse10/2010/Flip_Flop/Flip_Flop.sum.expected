-------------------------------------------------------------------------------
                          Semantic Analysis Summary                            
                                POGS GPL 2010                                  
            Copyright (C) 2010 Altran Praxis Limited, Bath, U.K.               
-------------------------------------------------------------------------------

Summary of:

Verification Condition files (.vcg)
Simplified Verification Condition files (.siv)
Proof Logs (.plg)
Dead Path Conjecture files (.dpc)
Summary Dead Path files (.sdp)

"status" column keys:
    1st character:
        '-' - No VC
        'S' - No SIV
        'U' - Undischarged
        'E' - Proved by Examiner
        'I' - Proved by Simplifier by Inference
        'X' - Proved by Simplifier by Contradiction
        'P' - Proved by Simplifier using User Defined Proof Rules
        'C' - Proved by Checker
        'R' - Proved by Review
        'F' - VC is False
    2nd character:
        '-' - No DPC
        'S' - No SDP
        'U' - Unchecked
        'D' - Dead path
        'L' - Live path

in the directory:
/Users/chalin/wkspdirs/36/sireum/sireum-spark-examples/src/org/sireum/profile/spark/examples/fse10/2010/Flip_Flop

Summary produced: 14-SEP-2010 14:08:01.35

File /Users/chalin/wkspdirs/36/sireum/sireum-spark-examples/src/org/sireum/profile/spark/examples/fse10/2010/Flip_Flop/flip_flop/flip.vcg
procedure Flip_Flop.Flip

VCs generated 14-SEP-2010 14:08:00

VCs simplified 14-SEP-2010 14:08:00

VCs for procedure_flip :
 -----------------------------------------------------------------------------
| #   | From  | To                  | Proved By          | Dead Path | Status |
|-----------------------------------------------------------------------------
| 1   | start | rtc check @ 45      | Inference          | No DPC    |   I-   |
| 2   | start | rtc check @ 47      | Inference          | No DPC    |   I-   |
| 3   | start |    assert @ finish  | Inference          | No DPC    |   I-   |
| 4   | start |    assert @ finish  | Inference          | No DPC    |   I-   |
| 5   |       | refinement          | Examiner           | No DPC    |   E-   |
| 6   |       | refinement          | Undischarged       | No DPC    |   U-   |
 -----------------------------------------------------------------------------


File /Users/chalin/wkspdirs/36/sireum/sireum-spark-examples/src/org/sireum/profile/spark/examples/fse10/2010/Flip_Flop/flip_flop/read.vcg
function Flip_Flop.Read

VCs generated 14-SEP-2010 14:08:00

VCs simplified 14-SEP-2010 14:08:00

VCs for function_read :
 -----------------------------------------------------------------------------
| #   | From  | To                  | Proved By          | Dead Path | Status |
|-----------------------------------------------------------------------------
| 1   | start |    assert @ finish  | Inference          | No DPC    |   I-   |
| 2   |       | refinement          | Examiner           | No DPC    |   E-   |
| 3   |       | refinement          | Undischarged       | No DPC    |   U-   |
 -----------------------------------------------------------------------------


File /Users/chalin/wkspdirs/36/sireum/sireum-spark-examples/src/org/sireum/profile/spark/examples/fse10/2010/Flip_Flop/flip_flop_/ptf/represents.vcg
function PTF.Represents

VCs generated 14-SEP-2010 14:08:00

VCs simplified 14-SEP-2010 14:08:01

VCs for function_represents :
 -----------------------------------------------------------------------------
| #   | From  | To                  | Proved By          | Dead Path | Status |
|-----------------------------------------------------------------------------
| 1   | start |    assert @ finish  | Inference          | No DPC    |   I-   |
 -----------------------------------------------------------------------------


===============================================================================
Summary:

The following subprograms have undischarged VCs (excluding those proved false):

   1  /Users/chalin/wkspdirs/36/sireum/sireum-spark-examples/src/org/sireum/profile/spark/examples/fse10/2010/Flip_Flop/flip_flop/flip.vcg
   1  /Users/chalin/wkspdirs/36/sireum/sireum-spark-examples/src/org/sireum/profile/spark/examples/fse10/2010/Flip_Flop/flip_flop/read.vcg

Proof strategies used by subprograms
-------------------------------------------------------------------------
Total subprograms with at least one VC proved by examiner:              2
Total subprograms with at least one VC proved by simplifier:            3
Total subprograms with at least one VC proved by contradiction:         0
Total subprograms with at least one VC proved with user proof rule:     0
Total subprograms with at least one VC proved using checker:            0
Total subprograms with at least one VC discharged by review:            0

Maximum extent of strategies used for fully proved subprograms:
-------------------------------------------------------------------------
Total subprograms with proof completed by examiner:                     0
Total subprograms with proof completed by simplifier:                   1
Total subprograms with proof completed with user defined rules:         0
Total subprograms with proof completed by checker:                      0
Total subprograms with VCs discharged by review:                        0

Overall subprogram summary:
-------------------------------------------------------------------------
Total subprograms fully proved:                                         1
Total subprograms with at least one undischarged VC:                    2  <<<
Total subprograms with at least one false VC:                           0
                                                                    -----
Total subprograms for which VCs have been generated:                    3


ZombieScope Summary:
-------------------------------------------------------------------------
Total subprograms for which DPCs have been generated:                   0
Total number subprograms with dead paths found:                         0
Total number of dead paths found:                                       0


VC summary:
-------------------------------------------------------------------------
Note: U/R denotes where the Simplifier has proved VCs using one or more user-
defined proof rules.

Total VCs by type:                                       
                            -----------Proved By Or Using------------
                     Total  Examiner Simp(U/R)  Checker Review False Undiscgd
Assert or Post:          4      0      4              0      0      0       0
Precondition check:      0      0      0              0      0      0       0
Check statement:         0      0      0              0      0      0       0
Runtime check:           2      0      2              0      0      0       0
Refinement VCs:          4      2      0              0      0      0       2
Inheritance VCs:         0      0      0              0      0      0       0
===============================================================================
Totals:                 10      2      6              0      0      0       2  <<<
% Totals:                      20%    60%             0%     0%     0%     20% <<<
===================== End of Semantic Analysis Summary ========================
